common/src/main/java/io/netty/util/ResourceLeakDetector.java

```diff
@@ -403,39 +423,56 @@ public boolean close(T trackedObject) {
 
         @Override
         public String toString() {
-            if (creationRecord == null) {
+            if (head == null) {
                 return EMPTY_STRING;
             }
 
-            final Object[] array;
-            final int removedRecords;
-            if (lastRecords != null) {
-                synchronized (lastRecords) {
-                    array = lastRecords.toArray();
-                    removedRecords = this.removedRecords;
+            final String creationRecord;
+            final String[] array;
+            int idx = 0;
+            String last = null;
+            final int dropped;
+
+            synchronized (head) {
+                if (tail == null) {
+                    // Already closed
+                    return EMPTY_STRING;
+                }
+                dropped = droppedRecords;
+                creationRecord = head.toString();
+                array = new String[numRecords];
+                Record record = head.next;
+                while (record != null) {
+                    String recordStr = record.toString();
+                    if (last == null || !last.equals(recordStr)) {
+                        array[idx ++] = recordStr;
+                        last = recordStr;
+                    }
+                    record = record.next;
                 }
-            } else {
-                removedRecords = 0;
-                array = EMPTY_OBJECTS;
             }
 
+            int removed = idx > MAX_RECORDS ? idx - MAX_RECORDS : 0;
+
+            long discarded = removed + dropped;
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Cast one of the operands of this addition operation to a "long". [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS2184)